wolfcrypt/src/dsa.c

```diff
@@ -654,263 +654,471 @@ int wc_DsaExportKeyRaw(DsaKey* dsa, byte* x, word32* xSz, byte* y, word32* ySz)
+    mp_int  *k = (mp_int *)XMALLOC(sizeof *k,
+                                   key->heap, DYNAMIC_TYPE_TMP_BUFFER);
+    mp_int  *kInv = (mp_int *)XMALLOC(sizeof *kInv,
+                                   key->heap, DYNAMIC_TYPE_TMP_BUFFER);
+    mp_int  *r = (mp_int *)XMALLOC(sizeof *r,
+                                   key->heap, DYNAMIC_TYPE_TMP_BUFFER);
+    mp_int  *s = (mp_int *)XMALLOC(sizeof *s,
+                                   key->heap, DYNAMIC_TYPE_TMP_BUFFER);
+    mp_int  *H = (mp_int *)XMALLOC(sizeof *H,
+                                   key->heap, DYNAMIC_TYPE_TMP_BUFFER);
 #ifndef WOLFSSL_MP_INVMOD_CONSTANT_TIME
-    mp_int  b;
+    mp_int  *b = (mp_int *)XMALLOC(sizeof *b,
+                                   key->heap, DYNAMIC_TYPE_TMP_BUFFER);
+#endif
+    byte    *buffer = (byte *)XMALLOC(DSA_HALF_SIZE, key->heap,
+                                      DYNAMIC_TYPE_TMP_BUFFER);
+#else
+    mp_int  k[1], kInv[1], r[1], s[1], H[1];
+#ifndef WOLFSSL_MP_INVMOD_CONSTANT_TIME
+    mp_int  b[1];
 #endif
-    mp_int* qMinus1;
-    int     ret = 0, sz;
     byte    buffer[DSA_HALF_SIZE];
+#endif
+    mp_int* qMinus1;
+    int     ret = 0, sz = 0;
     byte*   tmp;  /* initial output pointer */
 
-    if (digest == NULL || out == NULL || key == NULL || rng == NULL) {
-        return BAD_FUNC_ARG;
-    }
-
-    tmp = out;
-
-    sz = min((int)sizeof(buffer), mp_unsigned_bin_size(&key->q));
-
 #ifdef WOLFSSL_MP_INVMOD_CONSTANT_TIME
-    if (mp_init_multi(&k, &kInv, &r, &s, &H, 0) != MP_OKAY)
+    if (mp_init_multi(k, kInv, r, s, H, 0) != MP_OKAY)
 #else
-    if (mp_init_multi(&k, &kInv, &r, &s, &H, &b) != MP_OKAY)
+    if (mp_init_multi(k, kInv, r, s, H, b) != MP_OKAY)
 #endif
     {
-        return MP_INIT_E;
+        ret = MP_INIT_E;
+        goto out;
+    }
+
+#ifdef WOLFSSL_SMALL_STACK
+    if ((k == NULL) ||
+        (kInv == NULL) ||
+        (r == NULL) ||
+        (s == NULL) ||
+        (H == NULL)
+#ifndef WOLFSSL_MP_INVMOD_CONSTANT_TIME
+        || (b == NULL)
+#endif
+        || (buffer == NULL)) {
+        ret = MEMORY_E;
+        goto out;
+    }
+#endif
+
+    if (digest == NULL || out == NULL || key == NULL || rng == NULL) {
+        ret = BAD_FUNC_ARG;
+        goto out;
     }
-    qMinus1 = &kInv;
+
+    sz = min(DSA_HALF_SIZE, mp_unsigned_bin_size(&key->q));
+    tmp = out;
+    qMinus1 = kInv;
 
     /* NIST FIPS 186-4: B.2.2
      * Per-Message Secret Number Generation by Testing Candidates
      * Generate k in range [1, q-1].
      *   Check that k is less than q-1: range [0, q-2].
      *   Add 1 to k: range [1, q-1].
      */
-    if (mp_sub_d(&key->q, 1, qMinus1))
+    if (mp_sub_d(&key->q, 1, qMinus1)) {
         ret = MP_SUB_E;
+        goto out;
+    }
 
-    if (ret == 0) {
-        do {
-            /* Step 4: generate k */
-            ret = wc_RNG_GenerateBlock(rng, buffer, sz);
+    do {
+        /* Step 4: generate k */
+        if ((ret = wc_RNG_GenerateBlock(rng, buffer, sz))) {
```

Comment:
Checking ret for 0 is useful for compliance with certain standards like MISRA.
Don't take this out unless you have to.

Reply:
The current code in that PR makes it unambiguously impossible for `ret` to have a nonzero value at that codepoint, so I'm concerned it will generate compiler warnings and/or human confusion if it's checked for zeroness there.  Can you look over the broader context/flow in `wc_DsaSign()` and comment further?

Reply:
My issue is that you changed the way the code was written.
Instead of checking (ret == 0) && ... you now have gotos.
MISRA and such don't allow gotos and I suspect, though I don't know, that the code was written to this maxim.

Reply:
`goto fail` anyone?  :-)

I'll push a MISRA-kosher refactor using `break` momentarily.

Reply:
(done)